modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/RenderedPageUtil.java

```diff
@@ -0,0 +1,159 @@
+import com.liferay.portal.events.ThemeServicePreAction;
+import com.liferay.portal.kernel.model.Layout;
+import com.liferay.portal.kernel.service.LayoutLocalServiceUtil;
+import com.liferay.portal.kernel.servlet.DummyHttpServletResponse;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portal.vulcan.dto.converter.DTOConverterContext;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * @author Jürgen Kappler
+ */
+public class RenderedPageUtil {
+
+	public static RenderedPage getRenderedPage(
+			DTOConverterContext dtoConverterContext, Layout layout,
+			LayoutPageTemplateEntryLocalService
+				layoutPageTemplateEntryLocalService,
+			Portal portal)
+		throws Exception {
+
+		ThemeDisplay themeDisplay = _getThemeDisplay(
+			dtoConverterContext.getHttpServletRequest(), layout);
+
+		LayoutPageTemplateEntry layoutPageTemplateEntry =
+			_getLayoutPageTemplateEntry(
+				layout, layoutPageTemplateEntryLocalService, portal);
+
+		LayoutPageTemplateEntry masterLayout = _getMasterLayout(
+			layout, layoutPageTemplateEntryLocalService);
+
+		return new RenderedPage() {
+			{
+				renderedPageURL = portal.getLayoutRelativeURL(
+					layout, themeDisplay);
+
+				setMasterPageId(
+					() -> {
+						if (masterLayout != null) {
+							return masterLayout.getLayoutPageTemplateEntryKey();
+						}
+
+						return null;
+					});
+
+				setMasterPageName(
+					() -> {
+						if (masterLayout != null) {
+							return masterLayout.getName();
+						}
+
+						return null;
+					});
+
+				setPageTemplateId(
+					() -> {
+						if (layoutPageTemplateEntry != null) {
+							return layoutPageTemplateEntry.
+								getLayoutPageTemplateEntryKey();
+						}
+
+						return null;
+					});
+
+				setPageTemplateName(
+					() -> {
+						if (layoutPageTemplateEntry != null) {
+							return layoutPageTemplateEntry.getName();
+						}
+
+						return null;
+					});
+			}
+		};
+	}
+
+	private static LayoutPageTemplateEntry _getLayoutPageTemplateEntry(
+		Layout layout,
+		LayoutPageTemplateEntryLocalService layoutPageTemplateEntryLocalService,
+		Portal portal) {
+
+		if (layout.getClassNameId() != portal.getClassNameId(
+				LayoutPageTemplateEntry.class)) {
+
+			return null;
+		}
+
+		return layoutPageTemplateEntryLocalService.fetchLayoutPageTemplateEntry(
+			layout.getClassPK());
+	}
+
+	private static LayoutPageTemplateEntry _getMasterLayout(
+		Layout layout,
+		LayoutPageTemplateEntryLocalService
+			layoutPageTemplateEntryLocalService) {
+
+		Layout masterLayout = LayoutLocalServiceUtil.fetchLayout(
+			layout.getMasterLayoutPlid());
+
+		if (masterLayout == null) {
+			return null;
+		}
+
+		return layoutPageTemplateEntryLocalService.fetchLayoutPageTemplateEntry(
+			masterLayout.getClassPK());
+	}
+
+	private static ThemeDisplay _getThemeDisplay(
```

Comment:
Looks like we are going to need a provider for ThemeDisplay in the future if we keep needing things like this. Just a thought out loud :)